File: src/java/com/distributedlife/mahjong/reference/hand/Hand.java
package com.distributedlife.mahjong.reference.hand;

import com.distributedlife.mahjong.reference.data.TileSet;
import org.apache.commons.lang3.builder.CompareToBuilder;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import java.util.ArrayList;
import java.util.List;

public class Hand implements Comparable<Hand> {
    private final String name;
    private long part1 = 0;
    private long part2 = 0;
    private long part3 = 0;
    private long part4 = 0;

    public Hand(String name, List<String> requiredTiles) {
        this.name = name;

        if (requiredTiles != null) {
            convertHandToBitFields(requiredTiles);
        }
    }

    public Hand(String name, Long part1, Long part2, Long part3, Long part4) {
        this.name = name;
        this.part1 = part1;
        this.part2 = part2;
        this.part3 = part3;
        this.part4 = part4;
    }

    public String getName() {
        return name;
    }

    public String toString() {
        List<String> tiles = new ArrayList<String>();

        for (TileSet.Tile bitTile : TileSet.BIT_TILES) {
            if ((part1 & bitTile.v) == bitTile.v) {
                tiles.add(TileSet.getNameForBitTile(bitTile));
            }
        }
        for (TileSet.Tile bitTile : TileSet.BIT_TILES) {
            if ((part2 & bitTile.v) == bitTile.v) {
                tiles.add(TileSet.getNameForBitTile(bitTile));
            }
        }
        for (TileSet.Tile bitTile : TileSet.BIT_TILES) {
            if ((part3 & bitTile.v) == bitTile.v) {
                tiles.add(TileSet.getNameForBitTile(bitTile));
            }
        }
        for (TileSet.Tile bitTile : TileSet.BIT_TILES) {
            if ((part4 & bitTile.v) == bitTile.v) {
                tiles.add(TileSet.getNameForBitTile(bitTile));
            }
        }

        return tiles.toString();
    }

    private void convertHandToBitFields(List<String> requiredTiles) {
        List<String> remainingRequiredTiles = new ArrayList<String>(requiredTiles);

        for (TileSet.Tile bitTile : TileSet.BIT_TILES) {
            String tile = TileSet.getNameForBitTile(bitTile);

            if (remainingRequiredTiles.contains(tile)) {
                part1 += bitTile.v;
                remainingRequiredTiles.remove(tile);
            }
            if (remainingRequiredTiles.contains(tile)) {
                part2 += bitTile.v;
                remainingRequiredTiles.remove(tile);
            }
            if (remainingRequiredTiles.contains(tile)) {
                part3 += bitTile.v;
                remainingRequiredTiles.remove(tile);
            }
            if (remainingRequiredTiles.contains(tile)) {
                part4 += bitTile.v;
                remainingRequiredTiles.remove(tile);
            }
        }
    }

    @Override
    public int compareTo(Hand rhs) {
        return new CompareToBuilder()
                .append(name, rhs.name)
                .append(part1, rhs.part1)
                .append(part2, rhs.part2)
                .append(part3, rhs.part3)
                .append(part4, rhs.part4)
                .toComparison();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null)
            return false;
        if (obj == this)
            return true;
        if (!(obj instanceof Hand))
            return false;

        Hand rhs = (Hand) obj;
        return new EqualsBuilder().
                append(name, rhs.name).
                append(part1, rhs.part1).
                append(part2, rhs.part2).
                append(part3, rhs.part3).
                append(part4, rhs.part4).
                isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 31).
                append(name).
                append(part1).
                append(part2).
                append(part3).
                append(part4).
                toHashCode();
    }

    public long getPart1() {
        return part1;
    }

    public long getPart2() {
        return part2;
    }

    public long getPart3() {
        return part3;
    }

    public long getPart4() {
        return part4;
    }

    public boolean isPartialMatch(Hand compareTo) {
        return isPartialMatch(compareTo.getPart1());
    }

    public boolean isPartialMatch(Long part1) {
        if ((this.part1 & part1) > 0L) {
            return true;
        }

        return false;
    }
}

File: lib/hand/hand.js
/**
 * Hand - Represents a complete Mahjong hand using bit-field compression
 */
import { BIT_TILES, getNameForBitTile } from '../data/tile-set.js';

export class Hand {
  constructor(name, requiredTiles = null, part1 = 0n, part2 = 0n, part3 = 0n, part4 = 0n) {
    this.name = name;
    this.part1 = part1;
    this.part2 = part2;
    this.part3 = part3;
    this.part4 = part4;

    if (requiredTiles) {
      this.convertHandToBitFields(requiredTiles);
    }
  }

  convertHandToBitFields(requiredTiles) {
    const remainingRequiredTiles = [...requiredTiles];

    for (const bitTile of BIT_TILES) {
      const tile = getNameForBitTile(bitTile);

      if (remainingRequiredTiles.includes(tile)) {
        this.part1 += bitTile;
        remainingRequiredTiles.splice(remainingRequiredTiles.indexOf(tile), 1);
      }
      if (remainingRequiredTiles.includes(tile)) {
        this.part2 += bitTile;
        remainingRequiredTiles.splice(remainingRequiredTiles.indexOf(tile), 1);
      }
      if (remainingRequiredTiles.includes(tile)) {
        this.part3 += bitTile;
        remainingRequiredTiles.splice(remainingRequiredTiles.indexOf(tile), 1);
      }
      if (remainingRequiredTiles.includes(tile)) {
        this.part4 += bitTile;
        remainingRequiredTiles.splice(remainingRequiredTiles.indexOf(tile), 1);
      }
    }
  }

  toString() {
    const tiles = [];

    for (const bitTile of BIT_TILES) {
      if ((this.part1 & bitTile) === bitTile) {
        tiles.push(getNameForBitTile(bitTile));
      }
    }
    for (const bitTile of BIT_TILES) {
      if ((this.part2 & bitTile) === bitTile) {
        tiles.push(getNameForBitTile(bitTile));
      }
    }
    for (const bitTile of BIT_TILES) {
      if ((this.part3 & bitTile) === bitTile) {
        tiles.push(getNameForBitTile(bitTile));
      }
    }
    for (const bitTile of BIT_TILES) {
      if ((this.part4 & bitTile) === bitTile) {
        tiles.push(getNameForBitTile(bitTile));
      }
    }

    return tiles.toString();
  }

  compareTo(rhs) {
    if (this.name !== rhs.name) {
      return this.name < rhs.name ? -1 : 1;
    }
    if (this.part1 !== rhs.part1) {
      return this.part1 < rhs.part1 ? -1 : 1;
    }
    if (this.part2 !== rhs.part2) {
      return this.part2 < rhs.part2 ? -1 : 1;
    }
    if (this.part3 !== rhs.part3) {
      return this.part3 < rhs.part3 ? -1 : 1;
    }
    if (this.part4 !== rhs.part4) {
      return this.part4 < rhs.part4 ? -1 : 1;
    }
    return 0;
  }

  equals(obj) {
    if (!obj || !(obj instanceof Hand)) {
      return false;
    }
    return this.name === obj.name &&
           this.part1 === obj.part1 &&
           this.part2 === obj.part2 &&
           this.part3 === obj.part3 &&
           this.part4 === obj.part4;
  }

  hashCode() {
    let hash = 17n;
    hash = hash * 31n + BigInt(this.name.split('').reduce((acc, c) => acc + c.charCodeAt(0), 0));
    hash = hash * 31n + this.part1;
    hash = hash * 31n + this.part2;
    hash = hash * 31n + this.part3;
    hash = hash * 31n + this.part4;
    return hash;
  }

  isPartialMatch(compareTo) {
    const part1 = typeof compareTo === 'bigint' ? compareTo : compareTo.part1;
    return (this.part1 & part1) > 0n;
  }

  getName() {
    return this.name;
  }

  getPart1() {
    return this.part1;
  }

  getPart2() {
    return this.part2;
  }

  getPart3() {
    return this.part3;
  }

  getPart4() {
    return this.part4;
  }
}
